File: leetcode/js/finalPrices.js
/**
 * @param {number[]} prices
 * @return {number[]}
 */
const finalPrices = function (prices) {
  const stack = [];

  for (let i = 0; i < prices.length; i++) {
    while (stack.length && prices[stack[stack.length - 1]] >= prices[i]) {
      prices[stack.pop()] -= prices[i];
    }
    stack.push(i);
  }
  return prices;
};

console.log(finalPrices([8, 4, 6, 2, 3]));
console.log(finalPrices([1, 2, 3, 4, 5]));
console.log(finalPrices([10, 1, 1, 6]));

File: leetcode/python/finalPrices.py
from typing import List


class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        stack = []
        for ind, num in enumerate(prices):
            while stack and prices[stack[-1]] >= num:
                prices[stack.pop()] -= num
            stack.append(ind)

        return prices


print(Solution().finalPrices([8, 4, 6, 2, 3]))
